src/sandstorm/supervisor.c++

```diff
@@ -1541,6 +1541,35 @@ void SupervisorMain::DefaultSystemConnector::checkIfAlreadyRunning() const {
   KJ_UNREACHABLE;
 };
 
+typedef capnp::RealmGateway<capnp::Data, capnp::AnyPointer, ApiTokenOwner,
+                            capnp::AnyPointer> SupervisorRealmGateway;
+
+class SupervisorRealmGatewayImpl final: public SupervisorRealmGateway::Server {
+public:
+  explicit SupervisorRealmGatewayImpl(SandstormCore::Client&& sandstormCore)
+    : sandstormCore(kj::mv(sandstormCore)) {}
+
+  kj::Promise<void> import(ImportContext context) override {
+    auto cap = context.getParams().getCap().castAs<AppPersistent<> >();
+    auto owner = context.getParams().getParams().getSealFor();
+    return cap.saveRequest().send()
+        .then([this, owner, context](auto response) mutable {
+      auto req = sandstormCore.makeTokenRequest();
+      req.getRef().setAppRef(response.getObjectId());
+      req.setOwner(owner);
+      return req.send().then([context](auto response) mutable {
+        context.getResults().setSturdyRef(response.getToken());
+      });
+     });
+  }
+
+  kj::Promise<void> export_(ExportContext context) override {
+    KJ_FAIL_REQUIRE("Cannot call save() on capabilities across realms. Use SandstormApi.save().");
```

Comment:
Text: "Cannot directly call save() on capabilities outside the grain. Use SandstormApi.save() instead."